week-4/BookLibrary/Library.sol

```diff
@@ -0,0 +1,153 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.8.17;
+
+import "./BookRepository.sol";
+
+// Library contract is the main contract for the library
+contract Library is BookRepository {
+    // Membership fee for opening account
+    uint256 fee = 5 ether;
+
+    // Books that are borrowed by a member
+    struct Books {
+        uint256 bookId;
+        uint256 dueDate;
+    }
+
+    // Member struct to store member details
+    struct Member {
+        address payable memberId;
+        uint256 balance;
+        Books[] books;
+        bool isActive;
+    }
+
+    // memberRecords mapping to store member details
+    mapping(address => Member) public memberRecords;
+
+    // open account for new member
+    function openAccount() public payable {
+        // checks before opening account
+        require(
+            memberRecords[msg.sender].memberId == address(0),
+            "Account already exist"
+        );
+
+        require(
+            msg.value == fee,
+            "Pay 5 ETH membership fee for account opening!"
+        );
+
+        // add new member to memberRecords mapping
+        Member storage newMember = memberRecords[msg.sender];
```

Comment:
Creating a memory variable, updating all fields, and copying to the storage variable will be gas efficient.